### PR TITLE
Generate ops with type annotations

### PR DESCRIPTION
**Pull request by rahul-kamat**
Tensorflow Python APIs are generated in C++ and have Python wrappers as the public API. This PR will add type annotations to the generated python bindings of TensorFlow op kernels.

The changes in this PR do the following:
- Generate Type variables above each op with the accepted dtypes
- Generate type annotations for function arguments
- Generate type annotations for the tensor returned by the op

### Example
An op `Foo` is registed as:
```C++
REGISTER_OP("Foo”)
    .Input("x: T")
    .Output("y: Tout")
    .Attr("T: {int8, uint8}")
    .Attr("Tout: {complex64, complex128}")
```
The tensorflow python op generated with type annotations:
```python
TV_Foo_T = TypeVar("TV_Foo_T", dtypes.Int8, dtypes.UInt8)
TV_Foo_Tout = TypeVar("TV_Foo_Tout", dtypes.Complex64, dtypes.Complex128)

def foo(x: ops.Tensor[TV_Foo_T]) -> ops.Tensor[TV_Foo_Tout]:
   ...
```

Without these changes, the function definition would not have type annotations and will be generated as:
```python
def foo(x):
   ...
```

### CODE REVIEW

File: tensorflow/python/framework/python_op_gen.cc [hunk 1]
```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
...
+    if (attr.type() == "type") {
+      bool has_dtype_half = false;
+      for (int t : attr.allowed_values().list().type()) {
+        if (t == 19) { // DT_HALF = 19
+          has_dtype_half = true;
+          break;
+        }
+      }
+
+      // Do not add type annotations to an arg that can be of type half
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:422]: Let's write this as a // TODO: remove this once half is supported.

File: tensorflow/python/framework/python_op_gen.cc [hunk 2]
```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
...
+      if (has_dtype_half) continue;
+
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
+      param_type_map[attr.name()] = type_var_name;
+    }
+  }
+
+  for (int i = 0; i < api_def_.arg_order_size(); ++i) {
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:430]: Does this need to be done in arg order? Could we just enumerate over op_def_.input_arg instead?

File: tensorflow/python/framework/python_op_gen.cc [hunk 3]
```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
...
+
+      std::sort(allowed_types.begin(), allowed_types.end());
+
+      string typevar_dtypes;
+      for (std::vector<string>::iterator it = allowed_types.begin(); it != allowed_types.end(); ++it) {
+        if (!typevar_dtypes.empty()) strings::StrAppend(&typevar_dtypes, ", ");
+        strings::StrAppend(&typevar_dtypes, *it);
+      }
+
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:503]: It seems that the type var name is generated twice - let's only generate it once, and keep a map from attribute name to type var name.

File: tensorflow/python/framework/python_op_gen.cc [hunk 4]
```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
...
+
+  if(added_typevar) strings::StrAppend(&result_, "\n");
+}
+
+void GenEagerPythonOp::AddReturnTypeAnnotation() {
+  string return_type = "";
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
+    // If the "type" field is set, the return Tensor has a single DataType
+    if (arg.type() != 0) {
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:517]: There seems to be some redundancy here as well. Could we instead have a function that extracts the annotation code? Something like:

strings::StrAppend(&return_type, TypeAnnotationFor(arg));

**Review by mdanatg — COMMENTED** [review]
Good start. Added a few comments aimed at reducing code duplication a little.

File: tensorflow/python/framework/python_op_gen.cc [hunk 5]
```diff
@@ -311,10 +344,15 @@ string GenEagerPythonOp::Code() {
   }
 
   string parameters;
-  for (const auto& param : params_no_default_) {
-    if (!parameters.empty()) strings::StrAppend(&parameters, ", ");
-    strings::StrAppend(&parameters, param.GetRenameTo());
+  if (type_annotate_ops.find(op_def_.name()) != type_annotate_ops.end()) {
+    strings::StrAppend(&parameters, GetTypeAnnotatedParams());
+  } else {
+    for (const auto& param : params_no_default_) {
```
[the remark this answers is not in the document]
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc]: Made this change in `40ef6a7` and created `types_map` to lookup the type annotation for each param.

File: tensorflow/python/framework/python_op_gen.cc [hunk 6]
```diff
@@ -311,10 +344,15 @@ string GenEagerPythonOp::Code() {
...
+      if (!parameters.empty()) strings::StrAppend(&parameters, ", ");
+      strings::StrAppend(&parameters, param.GetRenameTo());
+    }
   }
+
   string parameters_with_defaults = parameters;
   for (const auto& param_and_default : params_with_default_) {
```
[the remark this answers is not in the document]
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc:371]: `50763f6` has the changes to annotate these params

File: tensorflow/python/framework/python_op_gen.cc (hunk 4 above)
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc]: Changed this to use a map to get the type annotation which cleaned up this function a lot.

File: tensorflow/python/framework/python_op_gen.cc [hunk 7]
```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
   return prelude_ + result_;
 }
 
+string GenEagerPythonOp::GetTypeAnnotatedParams() {
+  // holds mappings from param name to its type annotation
+  std::unordered_map<string, string> param_type_map;
+  for (int i = 0; i<op_def_.attr_size(); ++i) {
```
[the remark this answers is not in the document]
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc]: Yeah we don't need the `i`, I changed this loop and a few others as well in `8e70fe4`

File: tensorflow/python/framework/python_op_gen.cc [hunk 8]
```diff
@@ -310,16 +348,43 @@ string GenEagerPythonOp::Code() {
...
+  }
+
   string parameters;
   for (const auto& param : params_no_default_) {
     if (!parameters.empty()) strings::StrAppend(&parameters, ", ");
     strings::StrAppend(&parameters, param.GetRenameTo());
+
+    // Add type annotations to param
+    if (type_annotations.find(param.GetName()) != type_annotations.end()) {
+      if (!type_annotations[param.GetName()].empty()) {
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:364]: We can simplify this even more by not adding it to the map if it's empty.

File: tensorflow/python/framework/python_op_gen.cc [hunk 9]
```diff
@@ -310,16 +348,43 @@ string GenEagerPythonOp::Code() {
...
   string parameters_with_defaults = parameters;
   for (const auto& param_and_default : params_with_default_) {
     if (!parameters.empty()) strings::StrAppend(&parameters, ", ");
     if (!parameters_with_defaults.empty())
       strings::StrAppend(&parameters_with_defaults, ", ");
+
+    // Add type annotations to param_and_default
+    if (type_annotations.find(param_and_default.first.GetName()) != type_annotations.end()) {
+      if (!type_annotations[param_and_default.first.GetName()].empty()) {
+        strings::StrAppend(&parameters, ": ", type_annotations[param_and_default.first.GetName()]);
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:379]: This entry appears to be duplicate - make sure you capture it in a test.

File: tensorflow/python/framework/python_op_gen.cc [hunk 10]
```diff
@@ -310,16 +348,43 @@ string GenEagerPythonOp::Code() {
...
+
   string parameters_with_defaults = parameters;
   for (const auto& param_and_default : params_with_default_) {
     if (!parameters.empty()) strings::StrAppend(&parameters, ", ");
     if (!parameters_with_defaults.empty())
       strings::StrAppend(&parameters_with_defaults, ", ");
+
+    // Add type annotations to param_and_default
+    if (type_annotations.find(param_and_default.first.GetName()) != type_annotations.end()) {
+      if (!type_annotations[param_and_default.first.GetName()].empty()) {
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:378]: Suggest using a param_name local to avoid repeated calls - it makes the code cleaner, too.

File: tensorflow/python/framework/python_op_gen.cc [hunk 11]
```diff
@@ -367,6 +432,108 @@ string GenEagerPythonOp::Code() {
...
+  }
+
+  // Mapping input Tensors to their types
+  for (const auto& arg : op_def_.input_arg()) {
+    // Do not add type annotations to args that accept a sequence of Tensors
+    if (!arg.number_attr().empty()) continue;
+    string type_annotation;
+    if (type_annotations.find(arg.type_attr()) != type_annotations.end()) {
+      // Get the correct TypeVar if input maps to an attr
+      strings::StrAppend(&type_annotation, "_ops.Tensor[", type_annotations[arg.type_attr()], "]");
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:455]: I think the StrAppends can be avoided (we only set one item at any time), and can write directly into the map?

File: tensorflow/python/framework/python_op_gen.cc [hunk 12]
```diff
@@ -367,6 +432,108 @@ string GenEagerPythonOp::Code() {
...
+        strings::StrAppend(&type_annotation, "_ops.Tensor[", dtype_type[py_dtype], "]");
+      }
+    }
+
+    type_annotations[arg.name()] = type_annotation;
+  }
+
+  // Mapping output Tensor to its type
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:469]: A helper function (doesn't need to be a method) that just gets called in both places could help avoid the repetition here.

Something like: `type_annotations[arg.name()] = ParamAnnotation(type_annotations, arg);`

File: tensorflow/python/framework/python_op_gen.cc [hunk 13]
```diff
@@ -367,6 +432,108 @@ string GenEagerPythonOp::Code() {
   return prelude_ + result_;
 }
 
+std::unordered_map<string, string> GenEagerPythonOp::GetTypeAnnotationMap() {
+  std::unordered_map<string, string> type_annotations;
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:436]: I recommend using two maps to avoid mixing attrs with inputs/outputs: for example, attr_types and input_output_types.

File: tensorflow/python/framework/python_op_gen.cc [hunk 14]
```diff
@@ -367,6 +432,108 @@ string GenEagerPythonOp::Code() {
...
+      std::vector<string> allowed_types;
+      for (int t : attr.allowed_values().list().type()) {
+        DataType dtype = static_cast<DataType>(t);
+        const string py_dtype = python_op_gen_internal::DataTypeToPython(dtype, "_dtypes.");
+        if (dtype_type.find(py_dtype) != dtype_type.end()) {
+          allowed_types.emplace_back(dtype_type[py_dtype]);
+        }
+      }
+
+      // If all dtypes are allowed, add them all
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:500]: Does this mean "If no dtypes specified, then all are allowed"?

File: tensorflow/python/framework/python_op_gen.cc [hunk 15]
```diff
@@ -367,6 +432,108 @@ string GenEagerPythonOp::Code() {
...
+  if (added_typevar) strings::StrAppend(&result_, "\n");
+}
+
+void GenEagerPythonOp::AddReturnTypeAnnotation(std::unordered_map<string, string>& type_annotations) {
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
+    // Add type annotations to param
+    if (type_annotations.find(arg.name()) != type_annotations.end()) {
+      if (!type_annotations[arg.name()].empty()) {
+        result_.erase(result_.length() - 2);
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:530]: Does this remove the ":\n" bit? Add a comment if so. Optional: can we insert our return annotation in the place where that colon is inserted, to avoid erasing (perhaps in AddDefLine)?

**Review by mdanatg — COMMENTED** [review]
Very nice! A few more finishing touches and we should be good to go.

File: tensorflow/python/framework/python_op_gen.cc (hunk 14 above)
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc:500]: Yeah exactly, an example of this is the cast op. The input tensor `x` will have no dtypes specified, so all dtypes all are allowed.

File: tensorflow/python/framework/python_op_gen.cc (hunk 13 above)
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc:436]: Resolving this since we spoke offline about how the op's parameters include both inputs and attrs.

File: tensorflow/python/framework/python_op_gen.cc (hunk 9 above)
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc]: Added a test to check for adding type annotations to default args in `8bec96c`.

File: tensorflow/python/framework/python_op_gen.cc [hunk 16]
```diff
@@ -355,18 +416,97 @@ string GenEagerPythonOp::Code() {
...
+      type_annotations[attr.name()] = type_var_name;
+    } else if (attr.type() == "bool" || attr.type() == "float" ||
+               attr.type() == "int" || attr.type() == "bytes") {
+      type_annotations[attr.name()] = attr.type();
+    }
+  }
+
+  // Mapping input Tensors to their types
+  for (const auto& arg : op_def_.input_arg()) {
+    // Do not add type annotations to args that accept a sequence of Tensors
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:446]: Let's make this a TODO (and it would make a nice addition in a separate PR, too).

File: tensorflow/python/framework/python_op_gen.cc [hunk 17]
```diff
@@ -355,18 +416,97 @@ string GenEagerPythonOp::Code() {
...
+    if (attr.type() == "type") {
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
+      type_annotations[attr.name()] = type_var_name;
+    } else if (attr.type() == "bool" || attr.type() == "float" ||
+               attr.type() == "int" || attr.type() == "bytes") {
+      type_annotations[attr.name()] = attr.type();
+    }
+  }
+
+  // Mapping input Tensors to their types
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:444]: Nit: Keep a consistent comment style - Map input Tensors ...

File: tensorflow/python/framework/python_op_gen.cc (hunk 14 above)
Comment (mdanatg) [reply tensorflow/python/framework/python_op_gen.cc:474]: SG. Please update the comment.

File: tensorflow/python/framework/python_op_gen.cc [hunk 18]
```diff
@@ -355,18 +416,97 @@ string GenEagerPythonOp::Code() {
...
+      added_typevar = true;
+    }
+  }
+
+  if (added_typevar) strings::StrAppend(&result_, "\n");
+}
+
+// TODO(rahulkamat): Modify AddDefLine() to add return type annotation
+void GenEagerPythonOp::AddReturnTypeAnnotation(std::unordered_map<string, string>& type_annotations) {
+  if (op_def_.output_arg_size() == 1) {
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:500]: Add a TODO to handle multiple return types.

File: tensorflow/python/framework/python_op_gen.cc [hunk 19]
```diff
@@ -355,18 +416,97 @@ string GenEagerPythonOp::Code() {
...
+      const string type_var_name = type_annotations[attr.name()];
+      strings::StrAppend(&result_, type_var_name, " = TypeVar(\"", type_var_name, "\", ", typevar_dtypes,")\n");
+      added_typevar = true;
+    }
+  }
+
+  if (added_typevar) strings::StrAppend(&result_, "\n");
+}
+
+// TODO(rahulkamat): Modify AddDefLine() to add return type annotation
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:498]: I think this todo is a bit out of context - it's better to move it next to the erase call, and explain that doing it would avoid having to erase from the string.

File: tensorflow/python/framework/python_op_gen.cc [hunk 20]
```diff
@@ -1090,4 +1245,19 @@ string GetPythonWrappers(const char* op_list_buf, size_t op_list_len) {
...
 }
 
+string GetArgAnnotation(const auto& arg, std::unordered_map<string, string>& type_annotations) {
+  if (type_annotations.find(arg.type_attr()) != type_annotations.end()) {
+    // Get the correct TypeVar if arg maps to an attr
+    return "_ops.Tensor[" + type_annotations[arg.type_attr()] + "]";
+  } else {
+    // Get the dtype of the Tensor
+    const string py_dtype = python_op_gen_internal::DataTypeToPython(arg.type(), "_dtypes.");
+    if (dtype_type.find(py_dtype) != dtype_type.end()) {
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen.cc:1255]: I think we expect to always find a matching dtype, correct?

File: tensorflow/python/framework/python_op_gen_test.cc [hunk 21]
```diff
@@ -20,22 +20,269 @@ limitations under the License.
...
 
   EXPECT_TRUE(absl::StrContains(code, "def case"));
-
   // TODO(mdan): Add tests to verify type annotations are correctly added.
 }
 
+TEST(PythonOpGen, TypeAnnotateSingleTypeTensor) {
+  OpList op_defs;
+  OpRegistry::Global()->Export(false, &op_defs);
+  protobuf::TextFormat::ParseFromString(kBaseOpDef, &op_defs);  // NOLINT
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen_test.cc:190]: What was the lint error here?

File: tensorflow/python/framework/python_op_gen_test.cc [hunk 22]
```diff
@@ -20,22 +20,269 @@ limitations under the License.
...
+  ApiDefMap api_def_map(op_defs);
+
+  string code = GetPythonOps(op_defs, api_def_map, {}, "", type_annotate_ops);
+
+  const string baz_def_line = "def baz(inputs, name=None):";
+
+  ExpectHasSubstr(code, baz_def_line);
+}
+
 // TODO(mdan): Include more tests with synhtetic ops and api defs.
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen_test.cc:283]: I think this is done now!

File: tensorflow/python/framework/python_op_gen_test.cc [hunk 23]
```diff
@@ -20,22 +20,269 @@ limitations under the License.
...
 
-  string code = GetPythonOps(ops, api_def_map, {}, "");
+  string code = GetPythonOps(ops, api_def_map, {}, "", {});
 
   EXPECT_TRUE(absl::StrContains(code, "def case"));
-
   // TODO(mdan): Add tests to verify type annotations are correctly added.
 }
 
+TEST(PythonOpGen, TypeAnnotateSingleTypeTensor) {
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen_test.cc:187]: These tests are good. Consider declaring the proto string inside the function function itself, so that one can read everything there is to know about the test without scrolling back and forth.

**Review by mdanatg — CHANGES_REQUESTED** [review]
Looks great. A few final edits remaining, most minor.

File: tensorflow/python/framework/python_op_gen.cc (hunk 20 above)
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc:1255]: Yeah, we will always find a match now that we added support for `dtypes.half`.

File: tensorflow/python/framework/python_op_gen_test.cc (hunk 23 above)
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen_test.cc:82]: This sounds good to me, I put the proto strings inside each test in `b00c93e`.

File: tensorflow/python/framework/python_op_gen_test.cc [hunk 24]
```diff
@@ -23,17 +23,437 @@ limitations under the License.
...
 
-  string code = GetPythonOps(ops, api_def_map, {}, "");
+  std::unordered_set<string> type_annotate_ops;
+  for (const auto& op : ops.op()) {
+    type_annotate_ops.insert(op.name());
+  }
+
+  string code = GetPythonOps(ops, api_def_map, {}, "", type_annotate_ops);
+
+  const string all_types = ", _dtypes.BFloat16, _dtypes.Bool, _dtypes.Complex128, _dtypes.Complex64, "
```
Comment (mdanatg) [inline tensorflow/python/framework/python_op_gen_test.cc:59]: Be careful about making these checks too exhaustive. We don't want modifications to the code to require updating this test too often.

**Review by mdanatg — DISMISSED** [review]

**Review by mdanatg — APPROVED** [review]

File: tensorflow/python/framework/python_op_gen.cc (hunk 1 above)
Comment (rahul-kamat) [reply tensorflow/python/framework/python_op_gen.cc]: Half will be supported with the changes in https://github.com/tensorflow/tensorflow/pull/40696

Comment (mdanatg) [reply tensorflow/python/framework/python_op_gen.cc]: I know that, but the idea is to make a note of that in the code. Otherwise future readers will think there is a special reason not to use these.
